File: nodejs/test/e2e/skills.test.ts
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *--------------------------------------------------------------------------------------------*/

import * as fs from "fs";
import * as os from "os";
import * as path from "path";
import { describe, expect, it, beforeAll, afterAll } from "vitest";
import { createSdkTestContext } from "./harness/sdkTestContext.js";

describe("Skills Configuration", async () => {
    const { copilotClient: client } = await createSdkTestContext();
    let skillsDir: string;
    const SKILL_MARKER = "PINEAPPLE_COCONUT_42";

    beforeAll(() => {
        // Create a temporary skills directory with a test skill
        skillsDir = fs.mkdtempSync(path.join(os.tmpdir(), "copilot-skills-test-"));

        // Create a skill subdirectory with SKILL.md
        const skillSubdir = path.join(skillsDir, "test-skill");
        fs.mkdirSync(skillSubdir, { recursive: true });

        // Create a skill that instructs the model to include a specific marker in responses
        const skillContent = `---
name: test-skill
description: A test skill that adds a marker to responses
---

# Test Skill Instructions

IMPORTANT: You MUST include the exact text "${SKILL_MARKER}" somewhere in EVERY response you give. This is a mandatory requirement. Include it naturally in your response.
`;
        fs.writeFileSync(path.join(skillSubdir, "SKILL.md"), skillContent);
    });

    afterAll(() => {
        // Clean up the temporary skills directory
        if (skillsDir && fs.existsSync(skillsDir)) {
            fs.rmSync(skillsDir, { recursive: true, force: true });
        }
    });

    describe("Skill Behavior", () => {
        it("should load and apply skill from skillDirectories", async () => {
            const session = await client.createSession({
                skillDirectories: [skillsDir],
            });

            expect(session.sessionId).toBeDefined();

            // The skill instructs the model to include a marker - verify it appears
            const message = await session.sendAndWait({
                prompt: "Say hello briefly.",
            });

            expect(message?.data.content).toContain(SKILL_MARKER);

            await session.destroy();
        });

        it("should not apply skill when disabled via disabledSkills", async () => {
            const session = await client.createSession({
                skillDirectories: [skillsDir],
                disabledSkills: ["test-skill"],
            });

            expect(session.sessionId).toBeDefined();

            // The skill is disabled, so the marker should NOT appear
            const message = await session.sendAndWait({
                prompt: "Say hello briefly.",
            });

            expect(message?.data.content).not.toContain(SKILL_MARKER);

            await session.destroy();
        });

        it("should apply skill on session resume with skillDirectories", async () => {
            // Create a session without skills first
            const session1 = await client.createSession();
            const sessionId = session1.sessionId;

            // First message without skill - marker should not appear
            const message1 = await session1.sendAndWait({ prompt: "Say hi." });
            expect(message1?.data.content).not.toContain(SKILL_MARKER);

            // Resume with skillDirectories - skill should now be active
            const session2 = await client.resumeSession(sessionId, {
                skillDirectories: [skillsDir],
            });

            expect(session2.sessionId).toBe(sessionId);

            // Now the skill should be applied
            const message2 = await session2.sendAndWait({
                prompt: "Say hello again.",
            });

            expect(message2?.data.content).toContain(SKILL_MARKER);

            await session2.destroy();
        });
    });

    describe("Multiple Skills", () => {
        it("should load skills from multiple directories", async () => {
            const SKILL2_MARKER = "MANGO_BANANA_99";

            // Create a second temporary skills directory
            const skillsDir2 = fs.mkdtempSync(path.join(os.tmpdir(), "copilot-skills-test2-"));
            const skillSubdir2 = path.join(skillsDir2, "test-skill-2");
            fs.mkdirSync(skillSubdir2, { recursive: true });

            const skillContent2 = `---
name: test-skill-2
description: Second test skill that adds another marker
---

# Second Skill Instructions

IMPORTANT: You MUST include the exact text "${SKILL2_MARKER}" somewhere in EVERY response. This is mandatory.
`;
            fs.writeFileSync(path.join(skillSubdir2, "SKILL.md"), skillContent2);

            try {
                const session = await client.createSession({
                    skillDirectories: [skillsDir, skillsDir2],
                });

                const message = await session.sendAndWait({
                    prompt: "Say something brief.",
                });

                // Both skill markers should appear
                expect(message?.data.content).toContain(SKILL_MARKER);
                expect(message?.data.content).toContain(SKILL2_MARKER);

                await session.destroy();
            } finally {
                fs.rmSync(skillsDir2, { recursive: true, force: true });
            }
        });
    });
});

File: go/e2e/skills_test.go
package e2e

import (
	"os"
	"path/filepath"
	"strings"
	"testing"
	"time"

	copilot "github.com/github/copilot-sdk/go"
	"github.com/github/copilot-sdk/go/e2e/testharness"
)

const skillMarker = "PINEAPPLE_COCONUT_42"

func createTestSkillDir(t *testing.T, marker string) string {
	skillsDir, err := os.MkdirTemp("", "copilot-skills-test-")
	if err != nil {
		t.Fatalf("Failed to create temp skills directory: %v", err)
	}

	skillSubdir := filepath.Join(skillsDir, "test-skill")
	if err := os.MkdirAll(skillSubdir, 0755); err != nil {
		t.Fatalf("Failed to create skill subdirectory: %v", err)
	}

	skillContent := `---
name: test-skill
description: A test skill that adds a marker to responses
---

# Test Skill Instructions

IMPORTANT: You MUST include the exact text "` + marker + `" somewhere in EVERY response you give. This is a mandatory requirement. Include it naturally in your response.
`
	if err := os.WriteFile(filepath.Join(skillSubdir, "SKILL.md"), []byte(skillContent), 0644); err != nil {
		t.Fatalf("Failed to write SKILL.md: %v", err)
	}

	return skillsDir
}

func TestSkillBehavior(t *testing.T) {
	ctx := testharness.NewTestContext(t)
	client := ctx.NewClient()
	t.Cleanup(func() { client.ForceStop() })

	skillsDir := createTestSkillDir(t, skillMarker)
	t.Cleanup(func() { os.RemoveAll(skillsDir) })

	t.Run("load and apply skill from skillDirectories", func(t *testing.T) {
		ctx.ConfigureForTest(t)

		session, err := client.CreateSession(&copilot.SessionConfig{
			SkillDirectories: []string{skillsDir},
		})
		if err != nil {
			t.Fatalf("Failed to create session: %v", err)
		}

		// The skill instructs the model to include a marker - verify it appears
		message, err := session.SendAndWait(copilot.MessageOptions{
			Prompt: "Say hello briefly.",
		}, 60*time.Second)
		if err != nil {
			t.Fatalf("Failed to send message: %v", err)
		}

		if message.Data.Content == nil || !strings.Contains(*message.Data.Content, skillMarker) {
			t.Errorf("Expected message to contain skill marker '%s', got: %v", skillMarker, message.Data.Content)
		}

		session.Destroy()
	})

	t.Run("not apply skill when disabled via disabledSkills", func(t *testing.T) {
		ctx.ConfigureForTest(t)

		session, err := client.CreateSession(&copilot.SessionConfig{
			SkillDirectories: []string{skillsDir},
			DisabledSkills:   []string{"test-skill"},
		})
		if err != nil {
			t.Fatalf("Failed to create session: %v", err)
		}

		// The skill is disabled, so the marker should NOT appear
		message, err := session.SendAndWait(copilot.MessageOptions{
			Prompt: "Say hello briefly.",
		}, 60*time.Second)
		if err != nil {
			t.Fatalf("Failed to send message: %v", err)
		}

		if message.Data.Content != nil && strings.Contains(*message.Data.Content, skillMarker) {
			t.Errorf("Expected message to NOT contain skill marker '%s' when disabled, got: %v", skillMarker, *message.Data.Content)
		}

		session.Destroy()
	})

	t.Run("apply skill on session resume with skillDirectories", func(t *testing.T) {
		ctx.ConfigureForTest(t)

		// Create a session without skills first
		session1, err := client.CreateSession(nil)
		if err != nil {
			t.Fatalf("Failed to create session: %v", err)
		}
		sessionID := session1.SessionID

		// First message without skill - marker should not appear
		message1, err := session1.SendAndWait(copilot.MessageOptions{Prompt: "Say hi."}, 60*time.Second)
		if err != nil {
			t.Fatalf("Failed to send message: %v", err)
		}

		if message1.Data.Content != nil && strings.Contains(*message1.Data.Content, skillMarker) {
			t.Errorf("Expected message to NOT contain skill marker before skill was added, got: %v", *message1.Data.Content)
		}

		// Resume with skillDirectories - skill should now be active
		session2, err := client.ResumeSessionWithOptions(sessionID, &copilot.ResumeSessionConfig{
			SkillDirectories: []string{skillsDir},
		})
		if err != nil {
			t.Fatalf("Failed to resume session: %v", err)
		}

		if session2.SessionID != sessionID {
			t.Errorf("Expected session ID %s, got %s", sessionID, session2.SessionID)
		}

		// Now the skill should be applied
		message2, err := session2.SendAndWait(copilot.MessageOptions{Prompt: "Say hello again."}, 60*time.Second)
		if err != nil {
			t.Fatalf("Failed to send message: %v", err)
		}

		if message2.Data.Content == nil || !strings.Contains(*message2.Data.Content, skillMarker) {
			t.Errorf("Expected message to contain skill marker '%s' after resume, got: %v", skillMarker, message2.Data.Content)
		}

		session2.Destroy()
	})
}

func TestMultipleSkills(t *testing.T) {
	ctx := testharness.NewTestContext(t)
	client := ctx.NewClient()
	t.Cleanup(func() { client.ForceStop() })

	const skill2Marker = "MANGO_BANANA_99"

	skillsDir := createTestSkillDir(t, skillMarker)
	t.Cleanup(func() { os.RemoveAll(skillsDir) })

	// Create a second skills directory
	skillsDir2, err := os.MkdirTemp("", "copilot-skills-test2-")
	if err != nil {
		t.Fatalf("Failed to create temp skills directory 2: %v", err)
	}
	t.Cleanup(func() { os.RemoveAll(skillsDir2) })

	skillSubdir2 := filepath.Join(skillsDir2, "test-skill-2")
	if err := os.MkdirAll(skillSubdir2, 0755); err != nil {
		t.Fatalf("Failed to create skill subdirectory 2: %v", err)
	}

	skillContent2 := `---
name: test-skill-2
description: Second test skill that adds another marker
---

# Second Skill Instructions

IMPORTANT: You MUST include the exact text "` + skill2Marker + `" somewhere in EVERY response. This is mandatory.
`
	if err := os.WriteFile(filepath.Join(skillSubdir2, "SKILL.md"), []byte(skillContent2), 0644); err != nil {
		t.Fatalf("Failed to write SKILL.md: %v", err)
	}

	t.Run("load skills from multiple directories", func(t *testing.T) {
		ctx.ConfigureForTest(t)

		session, err := client.CreateSession(&copilot.SessionConfig{
			SkillDirectories: []string{skillsDir, skillsDir2},
		})
		if err != nil {
			t.Fatalf("Failed to create session: %v", err)
		}

		message, err := session.SendAndWait(copilot.MessageOptions{
			Prompt: "Say something brief.",
		}, 60*time.Second)
		if err != nil {
			t.Fatalf("Failed to send message: %v", err)
		}

		// Both skill markers should appear
		if message.Data.Content == nil {
			t.Fatal("Expected non-nil message content")
		}
		if !strings.Contains(*message.Data.Content, skillMarker) {
			t.Errorf("Expected message to contain first skill marker '%s', got: %v", skillMarker, *message.Data.Content)
		}
		if !strings.Contains(*message.Data.Content, skill2Marker) {
			t.Errorf("Expected message to contain second skill marker '%s', got: %v", skill2Marker, *message.Data.Content)
		}

		session.Destroy()
	})
}
